### PR TITLE
Patch 6

### PR DESCRIPTION
**Pull request by cri-s**
small fixes/changes

### CODE REVIEW

File: src/main/java/org/openpnp/machine/reference/feeder/wizards/ReferenceTubeFeederConfigurationWizard.java [hunk 1]
```diff
@@ -19,13 +19,13 @@
 
 package org.openpnp.machine.reference.feeder.wizards;
 
-import org.openpnp.machine.reference.feeder.ReferenceTubeFeeder;
+import org.openpnp.machine.reference.ReferenceFeeder;
 
 public class ReferenceTubeFeederConfigurationWizard
         extends AbstractReferenceFeederConfigurationWizard {
-    private final ReferenceTubeFeeder feeder;
+    private final ReferenceFeeder feeder;
```
Comment (vonnieda) [inline src/main/java/org/openpnp/machine/reference/feeder/wizards/ReferenceTubeFeederConfigurationWizard.java]: What is the purpose of this change? It limits access to any feeder specific properties and this is not how any of the other wizards work.

File: src/main/java/org/openpnp/vision/pipeline/stages/Threshold.java [hunk 2]
```diff
@@ -45,6 +45,11 @@ public Result process(CvPipeline pipeline) throws Exception {
         Mat mat = pipeline.getWorkingImage();
         int type = invert ? Imgproc.THRESH_BINARY_INV : Imgproc.THRESH_BINARY;
         type |= auto ? Imgproc.THRESH_OTSU : 0;
+	int i,j;
+	i=threshold<-1000?-threshold%100:12;
+	j=threshold<-1000?-threshold/1000:2;
+	if(threshold<0)
+        Imgproc.adaptiveThreshold(mat, mat, 255, Imgproc.ADAPTIVE_THRESH_GAUSSIAN_C, type,i,j); else
```
Comment (vonnieda) [inline src/main/java/org/openpnp/vision/pipeline/stages/Threshold.java]: Can you please explain what this does? Some comments would be helpful. If the purpose is to allow for adaptive threshold this should probably be a new stage.

File: src/main/java/org/openpnp/machine/reference/feeder/wizards/ReferenceTubeFeederConfigurationWizard.java (hunk 1 above)
Comment (cri-s) [inline src/main/java/org/openpnp/machine/reference/feeder/wizards/ReferenceTubeFeederConfigurationWizard.java]: As TubeFeeder actually is the feeder with less features, basic feeder
wizard, it's used
for other feeders, this wizard could be used by other feeders that
need just minimal configurations.
Should i copy this as ReferenceFeederConfigurationWizard.java ? is this better ?

2016-07-26 17:28 GMT, Jason von Nieda <notifications@github.com>:

Comment (vonnieda) [inline src/main/java/org/openpnp/machine/reference/feeder/wizards/ReferenceTubeFeederConfigurationWizard.java]: AbstractReferenceFeederConfigurationWizard fulfills this purpose.

Comment (cri-s) [inline src/main/java/org/openpnp/machine/reference/feeder/wizards/ReferenceTubeFeederConfigurationWizard.java]: Adaptive threshold is only used sometimes for downlooking camera,
the usage is really limited and there are already many stages. I remove it.
If you want, i can update convertColor, but with FluentCv removed, i
post the code at the
end of this mail.

I have tried to use the abstract wizard, but got this compiling error:

[ERROR] Failed to execute goal
org.apache.maven.plugins:maven-compiler-plugin:3.1:compile
(default-compile) on project openpnp-gui: Compilation failure
[ERROR] /home/av/test/work/openpnp/openpnp-develop/x/openpnp-develop/new/openpnp-develop/src/main/java/org/openpnp/machine/reference/feeder/Bank.java:[223,17]
org.openpnp.machine.reference.feeder.wizards.AbstractReferenceFeederConfigurationWizard
is abstract; cannot be instantiated



----

package org.openpnp.vision.pipeline.stages;

import java.util.ArrayList;
import org.opencv.core.Mat;
import org.opencv.core.Core;
import org.opencv.imgproc.Imgproc;
import org.openpnp.vision.pipeline.CvPipeline;
import org.openpnp.vision.pipeline.CvStage;
import org.simpleframework.xml.Attribute;

/**
 * Converts the color of the current working image to the specified conversion.
 */
public class ConvertColor extends CvStage {


// copyed from fluentCV
    public enum ColorCode {
        Bgr2Gray(Imgproc.COLOR_BGR2GRAY),
        Bgr2Lum(-Imgproc.COLOR_BGR2Luv),
        Bgr2YCrCb(Imgproc.COLOR_BGR2YCrCb),
        YCrCb2Bgr(Imgproc.COLOR_YCrCb2BGR),
        Gray2Bgr(Imgproc.COLOR_GRAY2BGR),
        Bgr2Hls(Imgproc.COLOR_BGR2HLS),
        Hls2Bgr(Imgproc.COLOR_HLS2BGR),
        Bgr2HlsFull(Imgproc.COLOR_BGR2HLS_FULL),
        Hls2BgrFull(Imgproc.COLOR_HLS2BGR_FULL),
        Bgr2Hsv(Imgproc.COLOR_BGR2HSV),
        Hsv2Bgr(Imgproc.COLOR_HSV2BGR),
        Bgr2HsvFull(Imgproc.COLOR_BGR2HSV_FULL),
        Hsv2BgrFull(Imgproc.COLOR_HSV2BGR_FULL);

        private int code;

        ColorCode(int code) {
            this.code = code;
        }

        public int getCode() {
            return code;
        }
    }

    @Attribute
    private ColorCode conversion = ColorCode.Bgr2Gray;

    public ColorCode getConversion() {
        return conversion;
    }

    public void setConversion(ColorCode conversion) {
        this.conversion = conversion;
    }

    @Override
    public Result process(CvPipeline pipeline) throws Exception {
        Mat mat = pipeline.getWorkingImage();
	if(conversion.getCode()<0) {
		ArrayList<Mat> channels = new ArrayList<Mat>(3);
        	Imgproc.cvtColor(mat, mat, -conversion.getCode());
    		Core.split(mat, channels);
		mat.release(); mat=null;
		mat = channels.get(0);
	} else {
        	Imgproc.cvtColor(mat, mat, conversion.getCode());
	}
        return new Result(mat);
    }
}


2016-07-26 17:32 GMT, Cri S <phone.cri@gmail.com>:

Comment (vonnieda) [inline src/main/java/org/openpnp/machine/reference/feeder/wizards/ReferenceTubeFeederConfigurationWizard.java]: For adaptive, you can either remove it, create a new stage, or add a boolean that turns it on. Using a special number makes it hard to understand for users and the purpose of the pipeline stage system is to be self documenting and easy to understand. Easiest would just be to add a boolean property to turn it on and off. boolean adaptive = false. 

For the wizard, yes, the Abstract one is an abstract class, so it cannot be instantiated directly. You have to extend it but no additional code is required - just extend and provide a constructor. If you want a generic wizard you can extend it and call it ReferenceFeederConfigurationWizard

Comment (cri-s) [inline src/main/java/org/openpnp/machine/reference/feeder/wizards/ReferenceTubeFeederConfigurationWizard.java]: adaptive threshold was a temporary solution, i will submit better
threshold code later.

2016-07-26 17:59 GMT, Jason von Nieda <notifications@github.com>:

**Comment by vonnieda** [comment]
Thanks @cri-s!
